gnome-image-installer/pages/confirm/gis-confirm-page.c

```diff
@@ -0,0 +1,281 @@
+
+  priv->countdown_remaining = 30;
+  gis_confirm_page_countdown_cb (self);
+
+  priv->countdown_source = g_timeout_add_seconds (1,
+                                                  gis_confirm_page_countdown_cb,
+                                                  self);
+}
+
+static void
+gis_confirm_page_shown (GisPage *page)
+{
+  GisConfirmPage *self = GIS_CONFIRM_PAGE (page);
+  GisConfirmPagePrivate *priv = gis_confirm_page_get_instance_private (self);
+  GFile *image_file;
+  g_autofree gchar *image_basename = NULL;
+  UDisksClient *udisks_client;
+  UDisksBlock *target_block;
+  UDisksDrive *target_drive = NULL;
+  g_autofree gchar *target_model = NULL;
+  g_autofree gchar *target_size = NULL;
+  /* TODO */
+  gboolean target_computer_specified = TRUE;
+  gboolean target_computer_matches = TRUE;
+
+  if (gis_store_get_error() != NULL)
+    {
+      gis_confirm_page_advance (self);
+      return;
+    }
+
+  /* TODO: update DMI labels */
+
+  image_file = G_FILE (gis_store_get_object (GIS_STORE_IMAGE));
+  image_basename = g_file_get_basename (image_file);
+  gtk_label_set_text (priv->image1_label, image_basename);
+  udisks_client = UDISKS_CLIENT (gis_store_get_object (GIS_STORE_UDISKS_CLIENT));
+  target_block = UDISKS_BLOCK (gis_store_get_object (GIS_STORE_BLOCK_DEVICE));
+  target_drive = udisks_client_get_drive_for_block (udisks_client, target_block);
+
+  g_assert (target_drive != NULL);
+
+  target_model = g_strdup_printf ("%s %s",
+                                  udisks_drive_get_vendor (target_drive),
+                                  udisks_drive_get_model  (target_drive));
+  g_strstrip (target_model);
+  gtk_label_set_text (priv->model1_label, target_model);
+  gtk_label_set_text (priv->device1_label, udisks_block_get_device (target_block));
+  target_size = g_format_size (udisks_drive_get_size (target_drive));
+  gtk_label_set_text (priv->size1_label, target_size);
+
+  g_clear_object (&target_drive);
+
+  gtk_widget_set_visible (GTK_WIDGET (priv->countdown_label),
+                          target_computer_specified && target_computer_matches);
+  gtk_widget_set_visible (GTK_WIDGET (priv->wrong_computer_box),
+                          target_computer_specified && !target_computer_matches);
+  if (target_computer_specified && target_computer_matches)
+    {
+      gis_confirm_page_start_countdown (self);
+    }
+  else if (target_computer_specified && !target_computer_matches)
+    {
+      gtk_button_set_label (priv->continue_button,
+                            _("C_ontinue Anyway and Destroy Data"));
+    }
+
+  gtk_widget_grab_default (GTK_WIDGET (priv->cancel_button));
+  gtk_widget_grab_focus (GTK_WIDGET (priv->cancel_button));
+}
+
+static void
+gis_confirm_page_cancel_clicked (GtkButton *button,
+                                 GisConfirmPage *self)
+{
+  g_autoptr(GError) error =
+    g_error_new (G_IO_ERROR,
+                 G_IO_ERROR_CANCELLED,
+                 _("Unattended installation was cancelled."));
```

Comment:
Nitpick: This could be `g_error_new_literal()` (or event `const` and on the stack, although that’s a little ickier).